src/AppInstallerCLICore/VTSupport.cpp

```diff
@@ -17,92 +17,198 @@ namespace AppInstaller::CLI::VirtualTerminal
+                    // Don't log it in that case.
+                    LOG_LAST_ERROR_IF(GetLastError() != ERROR_INVALID_HANDLE);
+                }
+                else
+                {
+                    for (DWORD mode : modeModifierFallbacks)
                     {
-                        m_token = true;
-                        break;
+                        DWORD outMode = (previousMode & ~disabledFlags) | mode;
+                        if (!SetConsoleMode(hStd, outMode))
+                        {
+                            // Even if it is a different error, log it and try to carry on.
+                            LOG_LAST_ERROR_IF(GetLastError() != STATUS_INVALID_PARAMETER);
+                        }
+                        else
+                        {
+                            return true;
+                        }
                     }
                 }
             }
+
+            return false;
+        }
+
+        // Extracts a VT sequence, expected one of the form ESCAPE + prefix + result + suffix, returning the result part.
+        std::string ExtractSequence(std::istream& inStream, std::string_view prefix, std::string_view suffix)
+        {
+            std::string result;
+
+            if (inStream.peek() == AICLI_VT_ESCAPE[0])
+            {
+                result.resize(4095);
+                inStream.readsome(&result[0], result.size());
+                THROW_HR_IF(E_UNEXPECTED, static_cast<size_t>(inStream.gcount()) >= result.size());
+
+                result.resize(static_cast<size_t>(inStream.gcount()));
+
+                std::string_view resultView = result;
+                size_t overheadLength = 1 + prefix.length() + suffix.length();
+                if (resultView.length() <= overheadLength ||
+                    resultView.substr(1, prefix.length()) != prefix ||
+                    resultView.substr(resultView.length() - suffix.length()) != suffix)
+                {
+                    result.clear();
+                }
+                else
+                {
+                    result = result.substr(1 + prefix.length(), result.length() - overheadLength);
+                }
+            }
+
+            return result;
         }
     }
 
-    ConsoleModeRestore::~ConsoleModeRestore()
+    ConsoleModeRestoreBase::ConsoleModeRestoreBase(DWORD handle) : m_handle(handle) {}
+
+    ConsoleModeRestoreBase::~ConsoleModeRestoreBase()
     {
         if (m_token)
         {
-            LOG_LAST_ERROR_IF(!SetConsoleMode(GetStdHandle(STD_OUTPUT_HANDLE), m_previousMode));
+            LOG_LAST_ERROR_IF(!SetConsoleMode(GetStdHandle(m_handle), m_previousMode));
             m_token = false;
         }
     }
 
+    ConsoleModeRestore::ConsoleModeRestore() : ConsoleModeRestoreBase(STD_OUTPUT_HANDLE)
+    {
+        m_token = InitializeMode(STD_OUTPUT_HANDLE, m_previousMode, { ENABLE_VIRTUAL_TERMINAL_PROCESSING | DISABLE_NEWLINE_AUTO_RETURN, ENABLE_VIRTUAL_TERMINAL_PROCESSING });

```

Comment:
FWIW, I'd personally always recommend against using the old mode as a basis for setting the new mode. After all, why should a previous console application be allowed to determine what modes you want?